samza-sql-avro/src/main/java/org/apache/samza/sql/avro/AvroSamzaToRelMsgConverterFactory.java

```diff
@@ -0,0 +1,142 @@
+* under the License.
+*/
+
+package org.apache.samza.sql.avro;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.samza.SamzaException;
+import org.apache.samza.sql.data.SamzaSqlRelMessage;
+import org.apache.samza.sql.interfaces.SamzaToRelMsgConverter;
+import org.apache.samza.sql.interfaces.SamzaToRelMsgConverterFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This class converts a Samza Avro messages into Relational messages. This supports Samza messages where Key is a
+ * string and Value is an avro record.
+ *
+ * Most of the fields of the Avro Record maps with their corresponding relational types except for Arrays
+ * which are converted from Avro {@link org.apache.avro.generic.GenericData.Array} to {@link ArrayList}
+ *
+ * All the fields of the Avro record are represented as a column with their corresponding field name in the relational message.
+ * Key from the Samza message is represented as a special column {@link AvroSamzaToRelMsgConverterFactory#KEY_NAME}
+ */
+public class AvroSamzaToRelMsgConverterFactory implements SamzaToRelMsgConverterFactory {
+
+  public static final String KEY_NAME = "key";
+
+  /**
+   * Class that converts the avro field to their corresponding relational fields
+   * Array fields are converted from Avro {@link org.apache.avro.generic.GenericData.Array} to {@link ArrayList}
+   */
+  public enum RelTypeToObjConverter {
+
+    ArraySqlType {
+      @Override
+      Object convert(RelDataType relType, Object avroObj) {
+        ArrayList<Object> retVal = new ArrayList<>();
+        if (avroObj != null) {
+          retVal.addAll(((GenericData.Array) avroObj));
+        }
+        return retVal;
+      }
+    },
+
+    MapSqlType {
+      @Override
+      Object convert(RelDataType relType, Object obj) {
+        Map<String, Object> retVal = new HashMap<>();
+        if (obj != null) {
+          retVal.putAll((Map<String, ?>) obj);
+        }
+        return retVal;
+      }
+    },
+
+    RelRecordType {
+      @Override
+      Object convert(RelDataType relType, Object obj) {
+        return obj;
+      }
+    },
+
+    BasicSqlType {
+      @Override
+      Object convert(RelDataType relType, Object obj) {
+        return obj;
+      }
+    };
+
+    abstract Object convert(RelDataType relType, Object obj);
+  }
+
+  private static final Logger LOG = LoggerFactory.getLogger(AvroTypeFactoryImpl.class);
+
+  @Override
+  public SamzaToRelMsgConverter create(Properties config) {
+    return new AvroSamzaToRelMsgConverter(config);
+  }
+
+  private class AvroSamzaToRelMsgConverter implements SamzaToRelMsgConverter {
```

Comment:
@nickpan47 I am not entirely sure that is a good idea, think that will just constrain us without any benefits of compile time type safety, Because we are loading the SamzaToRelMsgConverter dynamically at runtime.